File: JanperCab/3-Application/Dtos/Invoice/InvoiceDto.cs
using _3_Application.Dtos.Customer;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace _3_Application.Dtos.Invoice
{
    public class InvoiceDto
    {
        public int Id { get; set; }

        public int EnquiryId { get; set; }

        public DateTime CreatedDate { get; set; }


        public int CustomerId { get; set; }

        public string CustomerReference { get; set; }


        public string DoorType { get; set; }

        public string DoorColor { get; set; }


        public string InvoiceTo { get; set; }

        public string InvoiceAddress { get; set; }

        public string InvoiceSuburb { get; set; }

        public string InvoiceState { get; set; }

        public string InvoicePostcode { get; set; }


        public string DeliveryTo { get; set; }

        public string DeliveryAddress { get; set; }

        public string DeliverySuburb { get; set; }

        public string DeliveryState { get; set; }

        public string DeliveryPostcode { get; set; }


        public bool HasFixedPrice { get; set; }

        public decimal GstRate { get; set; }

        public decimal DiscountRate { get; set; }

        public decimal DeliveryFee { get; set; }

        public decimal SubTotal { get; set; }

        public decimal TotalGst { get; set; }

        public decimal TotalPrice { get; set; }


        public CustomerDto Customer { get; set; }

        public ICollection<InvoiceComponentDto> InvoiceComponents { get; set; }

        public InvoiceDto()
        {
            InvoiceComponents = new Collection<InvoiceComponentDto>();
        }
    }
}

File: JanperCab/5-JanperCab/ClientApp/src/app/_models/invoice/InvoiceDto.ts
import { CabinetMakerDto } from './../customer/CabinetMakerDto';
import { CustomerDto } from 'src/app/_models/customer/CustomerDto';
import { Type } from 'class-transformer';
import { InvoiceComponentDto } from './InvoiceComponentDto';
import { DistributorDto } from '../customer/DistributorDto';
import { ManufacturerDto } from '../customer/ManufacturerDto';
export class InvoiceDto {
  id: number;
  enquiryId: number;
  createdDate: Date;

  customerId: number;
  customerReference: string;

  doorType: string;
  doorColor: string;

  invoiceTo: string;
  invoiceAddress: string;
  invoiceSuburb: string;
  invoiceState: string;
  invoicePostcode: string;

  deliveryTo: string;
  deliveryAddress: string;
  deliverySuburb: string;
  deliveryState: string;
  deliveryPostcode: string;

  hasFixedPrice: boolean;
  gstRate: number;
  discountRate: number;
  deliveryFee: number;
  subTotal: number;
  totalGst: number;
  totalPrice: number;

  @Type(() => CustomerDto, {
    keepDiscriminatorProperty: true,
    discriminator: {
      property: '$type',
      subTypes: [
        {
          value: CabinetMakerDto,
          name: '_3_Application.Dtos.Customer.CabinetMakerDto, 3-Application',
        },
        {
          value: DistributorDto,
          name: '_3_Application.Dtos.Customer.DistributorDto, 3-Application',
        },
        {
          value: ManufacturerDto,
          name: '_3_Application.Dtos.Customer.ManufacturerDto, 3-Application',
        },
      ],
    },
  })
  customer: CustomerDto;

  @Type(() => InvoiceComponentDto)
  invoiceComponents: InvoiceComponentDto[];
}
